### PR TITLE
Build Windows DLL's

### LINKED ISSUE #1391: Dynamic linking in MSVC
Dynamic linking options are provided via the CMake option `BUILD_SHARED_LIBS`. However, simply stating that a library should be compiled as SHARED isn't enough in MSVC. In MSVC, you need to use `__declspec(export)` when compiling the library and `__declspec(import)` when using it, in order to export the functions/methods that should be visible to users of the DLL.

As far as I can see, this isn't currently done (I've looked at LinearMath). This means the library can't be used as a DLL when compiling with MSVC since no import library files are written.

The solution would be to define a macro:
```
#ifdef _WIN32
#ifdef LinearMath_EXPORTS
#define LINEARMATH_API __declspec(dllexport)
#else
#define LINEARMATH_API __declspec(dllimport)
#endif
#else
#define LINEARMATH_API
#endif
```
and use this macro when declaring a function/method that should be accessible to the user (so practically everywhere).

(`LinearMath_EXPORTS` is a definition CMake adds when building a shared library on Windows.)

### PR DESCRIPTION
**Pull request by madebr**
Hello
- ~I reformatted the main CMakeList.txt + those in the src subfolder to use spaces as was done in #2221 (If this is undesired, let me know and I'll updated those)~
- I kept everything in uppercase and used hard tabs in the cmake files
- I also updated the minimum required CMake version to 3.0. CMake 3.0 is the oldest CMake version that is nicely documented on cmake.org. I believe this is a nice compromise. FYI 2.4.3 was released on 2006-07-31, 3.0 on 2014-06-10. 
- ~I have added a `.editorconfig` file that should [automatically set source formatting properties](https://editorconfig.org/)~
- Added support for `BUILD_SHARED_LIBS` on Windows. When building shared libraries, all symbols are automatically exported. This fixes #1391 #1570 #1794 
- ~I also fiddled a bit with the examples to be able to build these using shared libraries.~
- On MSVC, the libraries in the examples, extras and tests folders are built as a static library by default. I focused on the main bullet3 libraries here.
- Both static, shared and pybullet libraries are built and tested on Appveyor
- ~I cannot reproduce the LNK4197 warnings on appveyor on a local Visual Studio 2017. So I assume these are fixed by MSVC.~
(I did not test mingw)
- The cmake script exports targets that consumers can use. It allows the following:

```
find_package(Bullet CONFIG 2.89 REQUIRED)
add_executable(my_simulator ${SOURCES})
target_link_libraries(my_simulator PUBLIC Bullet::BulletDynamics Bullet::BulletCollision)
```
This will test whether bullet has version of at least 2.89 and if the Bullet library is of the same architecture (32 vs 64 bit).
It will also automatically set the correct defines for linking to static/shared libraries and single/double precision.

### CODE REVIEW

File: Extras/Serialize/BulletFileLoader/btBulletFile.cpp [hunk 1]
```diff
@@ -13,6 +13,7 @@ subject to the following restrictions:
 3. This notice may not be removed or altered from any source distribution.
 */
 
+#include "Bullet3Serialize/Bullet2FileLoader/bullet2FileLoaderApi.h"
```
Comment (erwincoumans) [inline Extras/Serialize/BulletFileLoader/btBulletFile.cpp]: This dependency on that header is not wanted, please remove it.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.4.3)
+cmake_minimum_required(VERSION 3.0)
```
Comment (erwincoumans) [inline CMakeLists.txt]: What feature(s) are used that don't work in 2.4.3? Just bumping up to 3.0 without reason doesn't make sense.

File: Extras/Serialize/makesdna/CMakeLists.txt [hunk 3]
```diff
@@ -22,30 +22,24 @@ SET (INC_FILES
...
 IF (CMAKE_CL_64)
-	# Output BulletDNA.c
-	ADD_CUSTOM_COMMAND(
-		OUTPUT ${BULLET_PHYSICS_SOURCE_DIR}/src/LinearMath/btSerializer64.cpp
-		COMMAND ${CMAKE_CFG_INTDIR}/makesdna ${BULLET_PHYSICS_SOURCE_DIR}/src/LinearMath/btSerializer64.cpp ${BULLET_PHYSICS_SOURCE_DIR}/Extras/Serialize/CommonSerialize/
-		DEPENDS makesdna
-	)
-SET(SRC  ${BULLET_PHYSICS_SOURCE_DIR}/src/LinearMath/btSerializer64.cpp)
-
+  SET(TARGET_FN "${BULLET_PHYSICS_SOURCE_DIR}/src/LinearMath/btSerializer64.cpp")
```
Comment (erwincoumans) [inline Extras/Serialize/makesdna/CMakeLists.txt]: why did you remove this custom command?

**Comment by erwincoumans** [comment]
Thanks for the PR. There are a lot of changes and I will need to review each of them, not sure if/when I get to this.

**Comment by madebr** [comment]
I've been thinking.
This pr is trying to solve two problems (poorly):
- the bullet3 library is unable to compile and link as a dynamic library on Windows; and
- the examples are very interdependent.

Therefore I propose the following:
- fix the bullet3 linking in this pr (allow static and dynamic libraries; and
- always compile the examples as static libraries.

**Comment by madebr** [comment]
@erwincoumans I greatly simplified the pr and limited myself to:
- adding static/shared support for MSVC on Windows
- adding support for exporting targets (for consumers of bullet)
- getting pybullet to build on appveyor

**Comment by madebr** [comment]
Some questions:
- Does anyone use `INTERNAL_CREATE_DISTRIBUTABLE_MSVC_PROJECTFILES`? CMake does not create distributable project files because it depends on the actual toolchain (visual studio version etc). (I would remove it)
- I propose adding a namespace prefix to `INSTALL_LIBS` (`BULLET_INSTALL_LIBS`) because consumers who use bullet3 might want to install their own libs.

File: Extras/Serialize/BulletFileLoader/btBulletFile.cpp (hunk 1 above)
Comment (madebr) [reply Extras/Serialize/BulletFileLoader/btBulletFile.cpp]: This is removed in my most recent force push.

File: BulletConfigVersion.cmake.in [hunk 4]
```diff
@@ -0,0 +1,18 @@
+#                                               -*- cmake -*-
+#
+#  BulletConfigVersion.cmake(.in)
```
Comment (stephenjust) [inline BulletConfigVersion.cmake.in]: Is there any particular reason that you're not generating this with `WRITE_BASIC_PACKAGE_VERSION_FILE`? You've upgraded to CMake 3.0 anyways, so this is available to you.

Comment (madebr) [reply BulletConfigVersion.cmake.in]: I introduce it later in the commit `Export Bullet::* CMake targets so users can simply link to them`

File: CMakeLists.txt [hunk 5]
```diff
@@ -403,19 +405,22 @@ IF(BUILD_UNIT_TESTS)
...
-ENDIF()
-LIST(APPEND BULLET_INSTALL_LIBRARIES BulletSoftBody BulletDynamics BulletCollision BulletInverseDynamics LinearMath Bullet3Common)
-SET(BULLET_USE_FILE ${BULLET_INSTALL_CMAKEDIR}/UseBullet.cmake)
-CONFIGURE_FILE(${CMAKE_CURRENT_SOURCE_DIR}/BulletConfig.cmake.in
-	${CMAKE_CURRENT_BINARY_DIR}/BulletConfig.cmake
-	@ONLY ESCAPE_QUOTES
-)
+SET(BULLET_INSTALL_USEFILE "${BULLET_INSTALL_CMAKEDIR}/UseBullet.cmake")
+INCLUDE(CMakePackageConfigHelpers)
+CONFIGURE_PACKAGE_CONFIG_FILE(BulletConfig.cmake.in "${CMAKE_CURRENT_BINARY_DIR}/BulletConfig.cmake"
```
Comment (madebr) [inline CMakeLists.txt]: @stephenjust Here

File: BulletConfigVersion.cmake.in (hunk 4 above)
Comment (madebr) [reply BulletConfigVersion.cmake.in]: Oh you're right, I'm using the `WRITE_BASIC_PACKAGE_VERSION_FILE` function but I forgot to remove `BulletConfigVersion.cmake.in`

Comment (madebr) [reply BulletConfigVersion.cmake.in]: Force pushed a fix.

**Comment by lerppana** [comment]
Hello!

Is there any status for this PR?

**Comment by madebr** [comment]
@lerppana 
This pr set has become huge, too huge for merging.
It reformats the code, simplifies things here and there, adds support for shared libraries on Windows and fixes some examples.
That's why I extracted the formating to pr #2643.

Is @erwincoumans still interested in this, and willing to review #2643?
Because I don't see any feedback, I don't know whether it is worth the effort..

**Comment by lerppana** [comment]
Bullet is a great library, it deserves up to date build system. I hope this initiative moves forward 😊

**Comment by jiangcaiyang** [comment]
I strongly advise merging this PR. but I need to try branch in @madebr first to test.

**Comment by jiangcaiyang** [comment]
Hello @madebr , I can't run
```cmake --build %buildFolder% --config release --target install```
Because MSBuild tells me that project file install.vcxproj doesn't exist.
```
用于 .NET Framework 的 Microsoft (R) 生成引擎版本 16.7.0+b89cb5fde
版权所有(C) Microsoft Corporation。保留所有权利。

MSBUILD : error MSB1009: 项目文件不存在。
开关:install.vcxproj
```

**Comment by madebr** [comment]
@jiangcaiyang 
Installation is disabled by default on non-makefile systems.
You need to pass `-DINSTALL_LIBS=ON` to cmake when configuring to enable it.
